ops/ccs-ops-misc/load_test/high_volume_suite.py

```diff
@@ -0,0 +1,376 @@
+"""High Volume Load test suite for BFD Server endpoints."""
+
+from typing import Dict, List
+
+from locust import events, tag, task
+from locust.env import Environment
+
+from common import data, db, validation
+from common.bfd_user_base import BFDUserBase
+from common.locust_utils import is_distributed, is_locust_master
+from common.url_path import create_url_path
+
+MASTER_BENE_IDS: List[str] = []
+MASTER_CONTRACT_DATA: List[Dict[str, str]] = []
+MASTER_HASHED_MBIS: List[str] = []
+
+
+@events.test_start.add_listener
+def _(environment: Environment, **kwargs):
+    if (
+        is_distributed(environment)
+        and is_locust_master(environment)
+        or not environment.parsed_options
+    ):
+        return
+
+    # See https://docs.locust.io/en/stable/extending-locust.html#test-data-management
+    # for Locust's documentation on the test data management pattern used here
+    global MASTER_BENE_IDS
+    MASTER_BENE_IDS = data.load_from_parsed_opts(
+        environment.parsed_options,
+        db.get_bene_ids,
+        use_table_sample=True,
+        data_type_name="bene_ids",
+    )
+
+    global MASTER_CONTRACT_DATA
+    MASTER_CONTRACT_DATA = data.load_from_parsed_opts(
+        environment.parsed_options,
+        db.get_contract_ids,
+        use_table_sample=True,
+        data_type_name="contract_data",
+    )
+
+    global MASTER_HASHED_MBIS
+    MASTER_HASHED_MBIS = data.load_from_parsed_opts(
+        environment.parsed_options,
+        db.get_hashed_mbis,
+        use_table_sample=True,
+        data_type_name="hashed_mbis",
+    )
```

Comment:
One thing we should probably think about is whether we want to load data _once_ or _per-`HighVolumeUser`_. 

If we load _once_ (as the high volume suite is now), we may want to think about distributing data in some manner that ensures each `HighVolumeUser` isn't just hitting the same endpoint (i.e. 2 `HighVolumeUser`s get a `list` of `bene_id`s and, since they're the same list, they start sending requests to the _same_ endpoints at the same time). I can think of one fairly naive, but _very_ easy solution and that is to [`shuffle()`](https://docs.python.org/3/library/random.html#random.shuffle) each `list` of data in `HighVolumeUser`'s `__init__()` _after_ copying, i.e.:

```python
self.bene_ids = MASTER_BENE_IDS.copy()
shuffle(self.bene_ids)
...
```

Shuffling should ensure that, although the _initial_ data is the same, different IDs are being used between `HighVolumeUser`s at all times. I guess we could also split data between `HighVolumeUser`s in some fashion, but that would probably get fairly complex for little benefit.

The other option is to load data per-`HighVolumeUser` during Locust's [`on_start()`](https://docs.locust.io/en/stable/writing-a-locustfile.html#on-start-and-on-stop-methods). With `TABLESAMPLE` this should all but guarantee each `HighVolumeUser` gets a distinct set of data. The downside to this would be that we incur some setup time per-`HighVolumeUser`, _but_ if we limited it to something relatively small (like 1000 rows, or perhaps base it upon runtime) then I don't believe this would be too bad at all.

Reply:
I had some code in the run tests script to split the data amongst the users, but not sure how easy itd be to do with how things are generated now. Shuffle would at least help a little with avoiding caching, assuming we pulled enough data that we arent reusing too many of the same bene ids.

Reply:
I think splitting the **global** (loaded once) data across `HighVolumeUser`s would be more complexity than it's worth; honestly, I don't think that'd get us anything that loading 1000 (we could scale this with `spawned_runtime`)  or so pieces of data per `HighVolumeUser` in `on_start()` wouldn't.

Personally, I'd lean toward trying to load small amounts of data per-`HighVolumeUser` first. If that's too slow, or if it's not what we want for some reason, then just load globally and shuffle. If that isn't getting us what we want, then I feel like it'd be worth the effort to explore properly splitting a larger dataset across multiple `HighVolumeUser`s.

Reply:
There may be some considerations around this based on how we wind up running the tests also.
It might not be too bad of an effort if we go with some small number of users per table sample and then just have it request another batch when it exhausts the current batch. (This could be terrible though! I haven't looked into exact implementation of that and if Locust would make it difficult.)

Reply:
I added the shuffle. Doesn't seem to have much of a perceptible impact on performance either way, at least not with a limited number of users (10), but seems fine to keep in.